cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1651,6 +1688,58 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
       task.wait();
     }
   }
+
+  if (stats_granularity_ == statistics_freq::STATISTICS_COLUMN) {
+    // need pages on host to create offset_indexes
+    thrust::host_vector<gpu::EncPage> h_pages = cudf::detail::make_host_vector_async(pages, stream);
+    stream.synchronize();
+
+    // add column and offset indexes to metadata
+    for (auto b = 0, r = 0; b < static_cast<size_type>(batch_list.size()); b++) {
+      auto const rnext   = r + batch_list[b];
+      auto curr_page_idx = chunks[r][0].first_page;
+      for (; r < rnext; r++) {
+        int p                 = rg_to_part[r];
+        int global_r          = global_rowgroup_base[p] + r - first_rg_in_part[p];
+        auto const& row_group = md->file(p).row_groups[global_r];
+        for (auto i = 0; i < num_columns; i++) {
+          gpu::EncColumnChunk const& ck = chunks[r][i];
+          auto const& column_chunk_meta = row_group.columns[i].meta_data;
+
+          // start transfer of the column index
+          std::vector<uint8_t> column_idx;
+          column_idx.resize(ck.column_index_size);
+          CUDF_CUDA_TRY(cudaMemcpyAsync(column_idx.data(),
+                                        ck.column_index_blob,
+                                        ck.column_index_size,
+                                        cudaMemcpyDeviceToHost,
+                                        stream.value()));
+
+          // calculate offsets while the column index is transfering
+          int64_t curr_pg_offset = column_chunk_meta.data_page_offset;
+
+          OffsetIndex offset_idx;
+          for (uint32_t pg = 0; pg < ck.num_pages; pg++) {
+            auto const& enc_page = h_pages[curr_page_idx++];
+
+            // skip dict pages
+            if (enc_page.page_type != PageType::DATA_PAGE) { continue; }
+
+            int32_t this_page_size = enc_page.hdr_size + enc_page.max_data_size;
+            // first_row_idx is relative to start of row group
+            PageLocation loc{curr_pg_offset, this_page_size, enc_page.start_row - ck.start_row};
+            offset_idx.page_locations.push_back(loc);
+            curr_pg_offset += this_page_size;
+          }
+
+          stream.synchronize();
```

Comment:
Do we ever hit a situation where this is a blip on the radar? It seems we have work we can do in parallel here, but it may simply not be worth it.

Reply:
Here's where I'm a little out of my depth.  Are you saying don't interleave the memcpy with the offset index calc because there's no benefit, or just that the sync is not necessary because the data will show up in the time it takes to execute the loop?

Reply:
I like that overlap because why not do something while you wait, but I'm asking more about the larger loop here. We loop over `batch_list.size()` and do the same thing including this memcopy. Would it make sense to toss a kernel in here to do the calculations and then pull it back to the host as a `device_uvector` or, as a great person once said, is the juice not worth the squeeze?

Reply:
Ah, yes, I went back and forth on that before I started.  In the end I chose to do all the offset work on the host because the location of the first page needs to be known.  I could calculate 0 based offsets for each chunk on the device and then transfer them back, but would then have to add the offset of the first data page anyway.  There would be less to memcpy over, but I figured one memcpy for all the EncPage data vs more small memcpys for the arrays containing offsets would be faster in the end.  I lack a feel for these things, so could be convinced to try the latter approach.

Reply:
I would leave it for now and add an issue to research it.